File: interviews/turing/best_sum.py
from functools import lru_cache
from itertools import combinations

def best_sum_with_restrictions(nums):

    nums = sorted(nums, reverse=True)  # Sort to simplify initial checks
    max_sum = 0
    best_combination = []

    # Try all combinations of numbers to find the best valid sum
    for r in range(1, len(nums) + 1):
        for combination in combinations(nums, r):
            current_sum = 0
            valid = True

            # Validate the combination by the rule
            for num in sorted(combination):
                if num > current_sum:
                    current_sum += num
                else:
                    valid = False
                    break

            # Check if it's the best sum so far
            if valid and current_sum > max_sum:
                max_sum = current_sum
                best_combination = combination

    return max_sum, best_combination

def best_sum_with_restrictions_memo(nums):
    # Remove duplicates and sort descending
    nums = sorted(set(nums), reverse=True)  

    # Memoization cache to avoid recomputation
    @lru_cache(None)
    def dfs(current_sum):
        best_sum = current_sum
        best_combination = []

        # Try all numbers as candidates
        for num in nums:
            if num > current_sum:  # Ensure restriction is followed
                # Recur with updated sum
                new_sum, new_combination = dfs(current_sum + num)
                if new_sum > best_sum:  # Update if a better sum is found
                    best_sum = new_sum
                    best_combination = [num] + new_combination

        return best_sum, best_combination

    # Start from sum 0 and explore all possibilities
    return dfs(0)


# Memoization cache to avoid recomputation
# @lru_cache(None)
def best_sum_with_restrictions_rec(nums, res = 0):
    if len(nums) == 0:
        return res, []
    # Remove duplicates and sort descending
    nums = sorted(set(nums), reverse=True)  

    best_sum = [res for _ in nums]
    best_combination = [[] for _ in nums]

    # Try all numbers as candidates
    counter = -1
    for num in nums:
        counter += 1
        if num > best_sum[counter]:  # Ensure restriction is followed
            # Recursion with updated sum
            new_sum, new_combination = best_sum_with_restrictions_rec([n for n in nums if n>num], best_sum[counter] + num)
            if new_sum > best_sum[counter]:  # Update if a better sum is found
                best_sum[counter] = new_sum
                best_combination[counter] = [num] + new_combination

    idx = best_sum.index(max(best_sum))
    return best_sum[idx], best_combination[idx]

# Example usage with edge cases
print("Best Sum:      %d, Chosen Numbers: %s " %      best_sum_with_restrictions( [5, 1, 4, 8, 2, 20] )) #35
print("Best Sum Memo: %d, Chosen Numbers: %s " % best_sum_with_restrictions_memo( [5, 1, 4, 8, 2, 20] ))
print("Best Sum Rec:  %s, Chosen Numbers: %s " %  best_sum_with_restrictions_rec( [5, 1, 4, 8, 2, 20] ))
print("Best Sum:      %d, Chosen Numbers: %s " %      best_sum_with_restrictions( [5, 1, 3, 7, 2, 10] )) #19
print("Best Sum Memo: %d, Chosen Numbers: %s " % best_sum_with_restrictions_memo( [5, 1, 3, 7, 2, 10] ))
print("Best Sum Rec:  %s, Chosen Numbers: %s " %  best_sum_with_restrictions_rec( [5, 1, 3, 7, 2, 10] ))
print("Best Sum:      %d, Chosen Numbers: %s " %      best_sum_with_restrictions( [1, 1, 1, 1, 1, 1] )) #1
print("Best Sum Memo: %d, Chosen Numbers: %s " % best_sum_with_restrictions_memo( [1, 1, 1, 1, 1, 1] ))
print("Best Sum Rec:  %s, Chosen Numbers: %s " %  best_sum_with_restrictions_rec( [1, 1, 1, 1, 1, 1] ))
print("Best Sum:      %d, Chosen Numbers: %s " %      best_sum_with_restrictions( [50, 20, 30, 10, 5] )) #95
print("Best Sum Memo: %d, Chosen Numbers: %s " % best_sum_with_restrictions_memo( [50, 20, 30, 10, 5] ))
print("Best Sum Rec:  %s, Chosen Numbers: %s " %  best_sum_with_restrictions_rec( [50, 20, 30, 10, 5] ))

File: interviews/turing/best_sum.go
package main

import (
	"fmt"
	"sort"
)

func main() {
	testCases := [][]int{
		{5, 1, 4, 8, 2, 20},       // Should return 35 → [5, 2, 8, 20]
		{5, 1, 3, 7, 2, 10, 5, 3}, // 19
		{3, 2, 7, 10},             // Should return 19 → [2, 7, 10]
		{1, 1, 1, 1, 1, 1},        // Should return 1 → [1]
		{20, 5, 4, 3, 2, 1},       // Should return 35 → [20, 5, 4, 3, 2, 1]
		{50, 20, 30, 10, 5},       // Should return 95 → [5, 10, 30, 50]
	}

	for _, nums := range testCases {
		fmt.Printf("\nInput: %v\n", nums)

		bestSum, chosenNumbers := bestSumWithRestrictions(nums)
		fmt.Printf("Best Sum:      %d Chosen Numbers: %v\n", bestSum, chosenNumbers)
		bestSum, chosenNumbers = bestSumWithRestrictionsMemo(nums)
		fmt.Printf("Best Sum Memo: %d Chosen Numbers: %v\n", bestSum, chosenNumbers)
		bestSum, chosenNumbers = bestSumWithRestrictionsTab(nums)
		fmt.Printf("Best Sum Tab:  %d Chosen Numbers: %v\n", bestSum, chosenNumbers)
	}
}

// Helper function for recursive calculation
func bestSumWithRestrictions(nums []int) (int, []int) {
	// Remove duplicates by using a map
	uniqueNums := make(map[int]bool)
	for _, num := range nums {
		uniqueNums[num] = true
	}

	// Convert map back to slice and sort in descending order
	var sortedNums []int
	for num := range uniqueNums {
		sortedNums = append(sortedNums, num)
	}
	sort.Sort(sort.Reverse(sort.IntSlice(sortedNums)))

	// Recursive helper function
	var dfs func(currentSum int) (int, []int)
	dfs = func(currentSum int) (int, []int) {
		bestSum := currentSum
		var bestCombination []int

		for _, num := range sortedNums {
			if num > currentSum {
				newSum, newCombination := dfs(currentSum + num)
				if newSum > bestSum {
					bestSum = newSum
					bestCombination = append([]int{num}, newCombination...)
				}
			}
		}
		return bestSum, bestCombination
	}

	return dfs(0)
}

// Optimized function using memoization
func bestSumWithRestrictionsMemo(nums []int) (int, []int) {
	// Remove duplicates using a map and sort in descending order
	uniqueNums := make(map[int]bool)
	for _, num := range nums {
		uniqueNums[num] = true
	}

	var sortedNums []int
	for num := range uniqueNums {
		sortedNums = append(sortedNums, num)
	}
	sort.Sort(sort.Reverse(sort.IntSlice(sortedNums)))

	// Memoization map to store results for sums
	memo := make(map[int](struct {
		sum         int
		combination []int
	}))

	// Recursive helper function with memoization
	var dfs func(int) (int, []int)
	dfs = func(currentSum int) (int, []int) {
		// Return memoized result if available
		if result, found := memo[currentSum]; found {
			return result.sum, result.combination
		}

		bestSum := currentSum
		var bestCombination []int

		// Try all numbers and recursively explore further
		for _, num := range sortedNums {
			if num > currentSum {
				newSum, newCombination := dfs(currentSum + num)
				if newSum > bestSum {
					bestSum = newSum
					bestCombination = append([]int{num}, newCombination...)
				}
			}
		}

		// Store the result in memo before returning
		memo[currentSum] = struct {
			sum         int
			combination []int
		}{bestSum, bestCombination}

		return bestSum, bestCombination
	}

	return dfs(0)
}

// Space-optimized version for sum maximization with constraints
func bestSumWithRestrictionsTab(nums []int) (int, []int) {
	// Remove duplicates and sort ascending
	uniqueNums := make(map[int]bool)
	for _, num := range nums {
		uniqueNums[num] = true
	}

	var sortedNums []int
	for num := range uniqueNums {
		sortedNums = append(sortedNums, num)
	}
	sort.Ints(sortedNums) // Ascending order for proper constraint handling

	// Track current valid sums and their combinations
	currentSums := []int{0}
	bestCombination := map[int][]int{
		0: {},
	}

	bestSum := 0

	// Dynamic state expansion
	for _, num := range sortedNums {
		newSums := []int{}
		for _, sum := range currentSums {
			if num > sum {
				newSum := sum + num
				if _, exists := bestCombination[newSum]; !exists || newSum > bestSum {
					newSums = append(newSums, newSum)
					bestCombination[newSum] = append([]int{}, bestCombination[sum]...)
					bestCombination[newSum] = append(bestCombination[newSum], num)
					if newSum > bestSum {
						bestSum = newSum
					}
				}
			}
		}
		currentSums = append(currentSums, newSums...)
	}

	return bestSum, bestCombination[bestSum]
}
